packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts

```diff
@@ -0,0 +1,131 @@
+import type { Event, SignInExperience, Profile } from '@logto/schemas';
+import { SignUpIdentifier, SignInMode, SignInIdentifier } from '@logto/schemas';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import type { IdentifierPayload } from '../types/guard.js';
+
+const forbiddenEventError = new RequestError({ code: 'auth.forbidden', status: 403 });
+
+const forbiddenIdentifierError = new RequestError({
+  code: 'user.sign_in_method_not_enabled',
+  status: 422,
+});
+
+export const signInModeValidation = (event: Event, { signInMode }: SignInExperience) => {
+  if (event === 'sign-in') {
+    assertThat(signInMode !== SignInMode.Register, forbiddenEventError);
+  }
+
+  if (event === 'register') {
+    assertThat(signInMode !== SignInMode.SignIn, forbiddenEventError);
+  }
+};
+
+export const identifierValidation = (
+  identifier: IdentifierPayload,
+  signInExperience: SignInExperience
+) => {
+  const { signIn, signUp } = signInExperience;
+
+  // Username Password Identifier
+  if ('username' in identifier) {
+    assertThat(
+      signIn.methods.some(
+        ({ identifier: method, password }) => method === SignInIdentifier.Username && password
+      ),
+      forbiddenIdentifierError
+    );
+
+    return;
+  }
+
+  // Email Identifier
+  if ('email' in identifier) {
+    assertThat(
+      // eslint-disable-next-line complexity
+      signIn.methods.some(({ identifier: method, password, verificationCode }) => {
+        if (method !== SignInIdentifier.Email) {
+          return false;
+        }
+
+        // Email Password Verification
+        if ('password' in identifier && !password) {
+          return false;
+        }
```

Comment:
Would there be a case that both `password` and `passcode` are in identitifier?

Reply:
guarded by the input type